src/Agent/NewRelic/Agent/Core/DataTransport/DataStreamingService.cs

```diff
@@ -3,20 +3,72 @@
 * SPDX-License-Identifier: Apache-2.0
 */
 using System;
-using System.Collections.Concurrent;
 using System.Threading;
 using System.Threading.Tasks;
 using Grpc.Core;
 using NewRelic.Agent.Core.Utilities;
 using NewRelic.Core.Logging;
 using NewRelic.Agent.Core.AgentHealth;
 using NewRelic.Agent.Configuration;
-using System.Text;
 using System.Linq;
 using NewRelic.Collections;
+using System.Collections.Concurrent;
 
 namespace NewRelic.Agent.Core.DataTransport
 {
+
+    public class ResponseStreamWrapper<TResponse>
+    {
+        public readonly int ConsumerID;
+        
+        private bool _isInvalid = false;
+        public bool IsInvalid => _streamCancellationToken.IsCancellationRequested || _isInvalid;
+
+
+        private readonly IAsyncStreamReader<TResponse> _responseStream;
+        private readonly CancellationToken _streamCancellationToken;
+
+        private Task<int> _task;
+
+        public ResponseStreamWrapper(int consumerID, IAsyncStreamReader<TResponse> responseStream, CancellationToken streamCancellationToken)
+        {
+            ConsumerID = consumerID;
+            _responseStream = responseStream;
+            _streamCancellationToken = streamCancellationToken;
+        }
+
+        private async Task<int> WaitForResponse()
+        {
+            var success = false;
+
+            try
+            {
+                success = await _responseStream.MoveNext(_streamCancellationToken);
+            }
+            catch (Exception ex)
+            {
+                success = false;
```

Comment:
This will likely prevent us from knowing what error occurred when dealing with the response stream. I don't know if this will hurt us in the long run, but so far these errors have lead to us either catching problems with our usage of the grpc library or when the response stream is cancelled. Are we intentionally suppressing this level of detail?